### PR TITLE
Fix `equals` contract for `DefaultProject` and wrappers

### LINKED ISSUE #30272: Broken equals() contract for LifecycleAwareProject
### Current Behavior

The equals contract of `Project::equals` is broken when comparing against a `LifecycleAwareProject`: `LifecycleAwareProject` is equal to its corresponding `DefaultProject` but not the other way around.

### Expected Behavior

`equals()` must be symmetric

### Context (optional)

We have a plugin applied on the root project with an extension that maintains a `Set<Project>`.
A plugin applied on subprojects checks if the target project is contained in that set.

As of Gradle 8.10, the set now contains `LifecycleAwareProject`s and a `contains(defaultProject)` check always fails.

Also identity equality is not reliably anymore because of `LifecycleAwareProject` (e.g. in Java code: `project.getParent() == project.getRootProject()`). This changed behavior is understandable but might be good to document it in the release notes.

### Steps to Reproduce

/build.gradle.kts
```kotlin
project.ext["subproject"] = subprojects.toList()[0]
```

/subproject/build.gradle.kts
```kotlin
val lifecycleAware = rootProject.ext["subproject"]!!
println("$lifecycleAware (${lifecycleAware.javaClass}")
println("$project (${project.javaClass}")

println(lifecycleAware == project)
println(project == lifecycleAware)
```

This prints 
```
> Configure project :subproject
project ':subproject' (class org.gradle.api.internal.project.LifecycleAwareProject_Decorated
project ':subproject' (class org.gradle.api.internal.project.DefaultProject_Decorated
true
false
```

### Gradle version

8.10

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

### PR DESCRIPTION
**Pull request by 6hundreds**
Fixes #30272 

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by 6hundreds** [comment]
@bot-gradle test this without pts on linux

File: subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java [hunk 1]
```diff
@@ -596,6 +596,24 @@ public int compareTo(Project otherProject) {
...
+    public int hashCode() {
+        return super.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object otherProject) {
+        if (otherProject == null) {
+            return false;
+        }
+        if (otherProject instanceof MutableStateAccessAwareProject) {
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java:609]: 🤔 It doesn't feel right for `DefaultProject` to know about `MutableStateAccessAwareProejct`

File: subprojects/core/src/test/groovy/org/gradle/api/internal/project/DefaultProjectTest.groovy [hunk 2]
```diff
@@ -990,6 +990,34 @@ def scriptMethod(Closure closure) {
         p.subprojects { assertLifecycleAwareWithReferrer(it, p) }
     }
 
+    def equalsContractForWrappers() {
```
Comment (bamboo) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/project/DefaultProjectTest.groovy:993]: 🧪 I think we should have integration tests as well, comparing projects from within the different configuration scopes (e.g. `subprojects`, `allprojects`, `rootProject`).

**Review by bamboo — CHANGES_REQUESTED** [review]

File: subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java (hunk 1 above)
Comment (6hundreds) [reply subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java:609]: Hm, I remember that we were even considering put the logic of `MutableStateAccessAwareProject` to `DefaultProject` when discussions about the implementation were going. So given that, and that both types belongs to the same module and even package(which is a bit overinflated, tho), for me it feels ok to have this connection.

Do you see a ways to fix `rawProject.equals(wrappedProject)` scenario without mentioning wrappedProject type in `DefaultProject`?

I mean, I see the ways with introducing some equality helpers and move the logic there (like `ProjectOrderingUtils`), but it doesn't feel conceptually different.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/project/DefaultProjectTest.groovy (hunk 2 above)
Comment (6hundreds) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/project/DefaultProjectTest.groovy:993]: Added a test

**Comment by 6hundreds** [comment]
Superseded by #30319

File: subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java (hunk 1 above)
Comment (6hundreds) [reply subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java:609]: @bamboo please continue the review in https://github.com/gradle/gradle/pull/30319

**Pull request closed without merging**